ktlint-ruleset-standard/src/main/kotlin/com/pinterest/ktlint/ruleset/standard/rules/NoSameLineStatementAsOfBlockStartOrEnd.kt

```diff
@@ -0,0 +1,90 @@
+package com.pinterest.ktlint.ruleset.standard.rules
+
+import com.pinterest.ktlint.rule.engine.core.api.ElementType
+import com.pinterest.ktlint.rule.engine.core.api.IndentConfig
+import com.pinterest.ktlint.rule.engine.core.api.editorconfig.EditorConfig
+import com.pinterest.ktlint.rule.engine.core.api.editorconfig.INDENT_SIZE_PROPERTY
+import com.pinterest.ktlint.rule.engine.core.api.editorconfig.INDENT_STYLE_PROPERTY
+import com.pinterest.ktlint.rule.engine.core.api.indent
+import com.pinterest.ktlint.rule.engine.core.api.nextCodeLeaf
+import com.pinterest.ktlint.rule.engine.core.api.noNewLineInClosedRange
+import com.pinterest.ktlint.rule.engine.core.api.prevCodeLeaf
+import com.pinterest.ktlint.rule.engine.core.api.upsertWhitespaceAfterMe
+import com.pinterest.ktlint.rule.engine.core.api.upsertWhitespaceBeforeMe
+import com.pinterest.ktlint.ruleset.standard.StandardRule
+import org.jetbrains.kotlin.com.intellij.lang.ASTNode
+
+public class NoSameLineStatementAsOfBlockStartOrEnd
+    : StandardRule(
+    "no-same-line-statement-as-of-block-start-or-end",
+    usesEditorConfigProperties = setOf(
+        INDENT_SIZE_PROPERTY,
+        INDENT_STYLE_PROPERTY,
+    )
+) {
+    private var indentConfig = IndentConfig.DEFAULT_INDENT_CONFIG
+
+    override fun beforeFirstNode(editorConfig: EditorConfig) {
+        indentConfig =
+            IndentConfig(
+                indentStyle = editorConfig[INDENT_STYLE_PROPERTY],
+                tabWidth = editorConfig[INDENT_SIZE_PROPERTY],
+            )
+    }
+
+    override fun beforeVisitChildNodes(
+        node: ASTNode,
+        autoCorrect: Boolean,
+        emit: (offset: Int, errorMessage: String, canBeAutoCorrected: Boolean) -> Unit,
+    ) {
+        if (node.elementType == ElementType.FUN) {
+            node.findChildByType(ElementType.BLOCK)?.let { blockAstNode ->
+                val lBrace = blockAstNode.findChildByType(ElementType.LBRACE)!!
+                val rBrace = blockAstNode.findChildByType(ElementType.RBRACE)!!
+                val lBraceNextCodeElement = lBrace.nextCodeLeaf()!!
+                val rBracePrevCodeElement = rBrace.prevCodeLeaf()!!
+                val text = blockAstNode.text
```

Comment:
I am not a fan of declaring a lot of variables upfront. Also the `!!` I try to avoid. Most often, I just constructs like
```
blockAstNode
    .findChildByType(ElementType.LBRACE)
   ?.let { lbrace ->
   }
```

Reply:
That will create two level of nesting as I am using `rBrace` and `rBracePrevCodeElement` both and hence two `let` calls will be required

Reply:
Yes, that why I most often use that construct. I think for the lbrace the let-construction should be used. Starting from that point you know that you are inside a BLOCK that started with a LBRACE. This also means that RBRACE should exist (provided that the code could be compiled). So for RBRACE the `!!` or a `requireNotNull` can be used.

Reply:
Used `requireNotNull`